pkg/controller/elasticsearch/driver/upgrade_deletion_strategy_test.go

```diff
@@ -0,0 +1,540 @@
+	result := make([]string, len(pods))
+	for i, pod := range pods {
+		result[i] = pod.Name
+	}
+	return result
+}
+
+func (t testPod) toPod() corev1.Pod {
+	pod := corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      t.name,
+			Namespace: testNamespace,
+		},
+	}
+	labels := map[string]string{}
+	label.NodeTypesMasterLabelName.Set(t.master, labels)
+	label.NodeTypesDataLabelName.Set(t.data, labels)
+	pod.Labels = labels
+	if t.healthy {
+		pod.Status = corev1.PodStatus{
+			Conditions: []corev1.PodCondition{
+				{
+					Type:   corev1.PodReady,
+					Status: corev1.ConditionTrue,
+				},
+				{
+					Type:   corev1.ContainersReady,
+					Status: corev1.ConditionTrue,
+				},
+			},
+		}
+	}
+	return pod
+}
+
+type testESState struct {
+	inCluster []string
+	green     bool
+	ESState
+}
+
+func (t *testESState) GreenHealth() (bool, error) {
+	return t.green, nil
+}
+
+func (t *testESState) NodesInCluster(nodeNames []string) (bool, error) {
+	for _, nodeName := range nodeNames {
+		for _, inClusterPods := range t.inCluster {
+			if nodeName == inClusterPods {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+func (t *testESState) GetClusterState() (*esclient.ClusterState, error) {
+	var cs esclient.ClusterState
+	err := json.Unmarshal([]byte(sampleClusterState), &cs)
+	return &cs, err
+}
+
+func TestDeletionStrategy_Predicates(t *testing.T) {
+	type fields struct {
+		upgradeTestPods upgradeTestPods
+		ES              v1alpha1.Elasticsearch
+		green           bool
+	}
+	type args struct {
+		maxUnavailableReached bool
+		allowedDeletions      int
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		deleted []string
+		wantErr bool
+	}{
+		{
+			name: "3 healthy masters, allow the deletion of 1",
+			fields: fields{
+				upgradeTestPods: newUpgradeTestPods(
+					testPod{"masters-2", true, true, true, true, true},
+					testPod{"masters-1", true, true, true, true, true},
+					testPod{"masters-0", true, true, true, true, true},
+				),
+				green: true,
+			},
+			args: args{
+				maxUnavailableReached: false,
+				allowedDeletions:      1,
+			},
+			deleted: []string{"masters-2"},
+			wantErr: false,
+		},
+		{
+			name: "2 healthy masters out of 3, do not allow deletion",
+			fields: fields{
+				upgradeTestPods: newUpgradeTestPods(
+					testPod{"masters-0", true, true, true, true, true},
+					testPod{"masters-1", true, true, true, true, true},
+					testPod{"masters-2", true, true, false, false, false},
+				),
+
+				green: true,
+			},
+			args: args{
+				maxUnavailableReached: false,
+				allowedDeletions:      1,
+			},
+			deleted: []string{},
+			wantErr: false,
+		},
+		{
+			name: "1 master and 1 node, wait for the node to be upgraded first",
+			fields: fields{
+				upgradeTestPods: newUpgradeTestPods(
+					testPod{"masters-0", true, false, true, true, true},
+					testPod{"node-0", false, true, false, true, true},
+				),
+				green: true,
+			},
```

Comment:
Missing a test for the combined case single node master+data to upgrade (see my other comment)